### PR TITLE
Issue 2754 - Favourites endpoint

### LINKED ISSUE #2754: Favourites endpoint
### Description
New v5 endpoints to show favourite containers/federations.

#### Add to favourites
- Containers `PUT /teamspace/:teamspace/project/:project/containers/favourites`
```` json
{
  containers: [idA, idB, idC,]
}
````
This should add any ids that were not in the array to the list of favourites.

- Containers `PUT /teamspace/:teamspace/project/:project/federations/favourites`
```` json
{
  federations: [idA, idB, idC,]
}
````
This should add any ids that were not in the array to the list of favourites

#### Remove from favourites
- Containers `DELETE /teamspace/:teamspace/project/:project/containers/favourites`
```` json
{
  containers: [idA, idB, idC,]
}
````
This should remove any ids that are in the favourites array

- Containers `DELETE /teamspace/:teamspace/project/:project/federations/favourites`
```` json
{
  federations: [idA, idB, idC,]
}
````
This should remove any ids that are in the favourites array

NOTE: There is no need to have a GET endpoint for now as we have this embedded in the GET fed/container list (https://github.com/3drepo/3drepo.io/issues/2749

### Goals
- [ ] The user should be able to:
  - [ ] Add multiple ids into the favourites array
  - [ ] Remove multiple ids from the favourites array

### Tasks
- [ ] Implement endpoint
- [ ] Add unit tests
- [ ] Add e2e tests
- [ ] Add swagger doc

### PR DESCRIPTION
**Pull request by ChristosTsiotsias**
This fixes #2754 

#### Description
Favourites endpoints implemented along with the tests and documentation

#### Test cases
- One or more containers should be added to favourites list
- One or more federations should be added to favourites list
- One or more containers should be removed from favourites list
- One or more federations should be removed from favourites list
- Only users with adequate permissions on the models should be able to change the favourites list

### CODE REVIEW

File: backend/src/v5/routes/teamspaces/projects/containers/containers.js [hunk 1]
```diff
@@ -184,6 +203,94 @@ const establishRoutes = () => {
...
+	 *         schema:
+	 *           type: string
+	 *     requestBody:
+     *       content:
+     *         application/json:
+     *           schema:
+     *             type: object
+     *             properties:
+     *               containers:
+     *                 type: array
```
Comment (carmenfan) [inline backend/src/v5/routes/teamspaces/projects/containers/containers.js:235]: image

schema isn't complete here. you should define what the members of array should be (i.e. strings)

image

Also define an example so the doc reader knows what to put in

File: backend/src/v5/routes/teamspaces/projects/containers/containers.js [hunk 2]
```diff
@@ -184,6 +203,94 @@ const establishRoutes = () => {
...
+	 *     responses:
+	 *       401:
+	 *         $ref: "#/components/responses/notLoggedIn"
+	 *       404:
+	 *         $ref: "#/components/responses/teamspaceNotFound"
+	 *       200:
+	 *         description: adds the containers found in the request body to the user's favourites list
+	 *         content:
```
Comment (carmenfan) [inline backend/src/v5/routes/teamspaces/projects/containers/containers.js:243]: remove content, we don't need to send a response body if there's nothing to send.

File: backend/src/v5/routes/teamspaces/projects/containers/containers.js [hunk 3]
```diff
@@ -184,6 +203,94 @@ const establishRoutes = () => {
...
+	 *           application/json:
+	 *             schema:
+	 *               type: object
+	 */
+	router.patch('/favourites', hasAccessToTeamspace, appendFavourites);
+
+	/**
+	 * @openapi
+	 * /teamspaces/{teamspace}/projects/{project}/containers/favourites:
```
Comment (carmenfan) [inline backend/src/v5/routes/teamspaces/projects/containers/containers.js:252]: similar comments here

File: backend/src/v5/routes/teamspaces/projects/federations/federations.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+const Federations = require('../../../../processors/teamspaces/projects/federations/federations');
+const { Router } = require('express');
+const { hasAccessToTeamspace } = require('../../../../middleware/permissions/permissions');
+const { respond } = require('../../../../utils/responder');
+const { templates } = require('../../../../utils/responseCodes');
+
```
Comment (carmenfan) [inline backend/src/v5/routes/teamspaces/projects/federations/federations.js:23]: similar comments here

File: backend/tests/v5/e2e/routes/teamspaces/projects/containers/containers.test.js [hunk 5]
```diff
@@ -191,5 +224,6 @@ describe('E2E routes/teamspaces/projects/containers', () => {
 	});
 	afterAll(() => ServiceHelper.closeApp(server));
 	testGetContainerList();
+	testAppendFavourites();
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/teamspaces/projects/containers/containers.test.js:227]: missing remove favourites  tests?

File: backend/tests/v5/unit/middleware/permissions/teamspaces.test.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
+/**
```
Comment (carmenfan) [inline backend/tests/v5/unit/middleware/permissions/teamspaces.test.js:1]: again merge conflict error - this file should no longer exist

File: backend/tests/v5/unit/models/users.test.js [hunk 7]
```diff
@@ -72,7 +72,67 @@ const testGetFavourites = () => {
...
+	const favouritesData = {
+		teamspace1: ['a', 'b', 'c'],
+		teamspace2: ['d'],
+	};
+
+	describe('Add containers to favourites', () => {
+		test('Adding an array of containers of a new teamspace', async () => {
+			jest.spyOn(db, 'findOne').mockResolvedValue({ customData: { starredModels: favouritesData } });
+			await User.appendFavourites('xxx', 'teamspace3', ['e', 'f']);
+			const newFavourtiesData = favouritesData;
```
Comment (carmenfan) [inline backend/tests/v5/unit/models/users.test.js:85]: this test is meaningless - you are literally testing if the same object equals to the same object.

The only output you need to test is that it doesn't fail. so you don't need these 3 lines

Comment (carmenfan) [reply backend/tests/v5/unit/models/users.test.js:85]: same for the next 2 tests

File: backend/tests/v5/unit/models/users.test.js [hunk 8]
```diff
@@ -72,7 +72,67 @@ const testGetFavourites = () => {
...
+			newFavourtiesData.teamspace1 = ['d', 'e'];
+			expect(favouritesData).toEqual(newFavourtiesData);
+		});
+
+		test('Adding an array of containers of an existing teamspace with duplicate entries', async () => {
+			jest.spyOn(db, 'findOne').mockResolvedValue({ customData: { starredModels: favouritesData } });
+			await User.appendFavourites('xxx', 'teamspace1', ['a', 'b', 'c', ' d', 'e']);
+			const newFavourtiesData = favouritesData;
+			newFavourtiesData.teamspace1 = ['d', 'e'];
+			expect(favouritesData).toEqual(newFavourtiesData);
```
Comment (carmenfan) [inline backend/tests/v5/unit/models/users.test.js:103]: missing tests:
1. starredModels is undefined (this should append as normal and don't throw an error)
2. user doesn't exist (this should throw error)

File: backend/tests/v5/unit/models/users.test.js [hunk 9]
```diff
@@ -72,7 +72,67 @@ const testGetFavourites = () => {
...
+		});
+
+		test('Remove some containers of a teamspace', async () => {
+			jest.spyOn(db, 'findOne').mockResolvedValue({ customData: { starredModels: favouritesData } });
+			await User.deleteFavourites('xxx', 'teamspace1', ['c']);
+			const newFavourtiesData = favouritesData;
+			newFavourtiesData.teamspace1[2] = undefined;
+			expect(favouritesData).toEqual(newFavourtiesData);
+		});
```
Comment (carmenfan) [inline backend/tests/v5/unit/models/users.test.js:129]: missing tests:
1. user doesn't exist
2. teamspace doesn't have an entry
3. starredModels is undefined

File: backend/tests/v5/unit/processors/teamspaces/projects/containers/containers.test.js [hunk 10]
```diff
@@ -59,7 +59,7 @@ const containerSettings = {
 	},
 };
 
-const user1Favourites = [1];
+let user1Favourites = [1];
```
Comment (carmenfan) [inline backend/tests/v5/unit/processors/teamspaces/projects/containers/containers.test.js:62]: use a different dataset if this needs to change, so you don't contaminate across tests

File: backend/tests/v5/unit/processors/teamspaces/projects/containers/containers.test.js [hunk 11]
```diff
@@ -118,6 +134,62 @@ const testGetContainerList = () => {
 	});
 };
 
+const testAppendFavourites = () => {
+	describe('Add containers to favourites', () => {
+		test('user favourites should stay the same if one or more containers is not found', async () => {
+			await Containers.appendFavourites('tsAdmin', 'teamspace', 'project', [1, -1]);
+			expect(user1Favourites).toEqual([1]);
```
Comment (carmenfan) [inline backend/tests/v5/unit/processors/teamspaces/projects/containers/containers.test.js:141]: again you're not really testing anything within the codebase, you're essentially testing the mockImplementation you defined in this file.

test these round trip logic in the e2e tests, where you're dealing with real data. The whole point of unit tests are to test that the precondition and postcondition matches.

so in this case, appendFavourites should just work if you put in valid data, so the test you should be testing is:

1. favourites should update fine if the user has access to all the models
2. should throw an error if any if the models you cannot add to to the favourites
3. deal with edge cases - what if the list is empty?
4. what if the functions this function call throws an error -> do we deal with them appropriately?

File: backend/tests/v5/unit/processors/teamspaces/projects/containers/containers.test.js [hunk 12]
```diff
@@ -118,6 +134,62 @@ const testGetContainerList = () => {
...
+			await Containers.appendFavourites('user1', 'teamspace', 'project', [1, 2]);
+			expect(user1Favourites).toEqual([1]);
+			user1Favourites = [1];
+		});
+	});
+};
+
+const testDeleteFavourites = () => {
+	describe('Remove containers from favourites', () => {
+		test('user favourites should stay the same if one or more containers are not already in', async () => {
```
Comment (carmenfan) [inline backend/tests/v5/unit/processors/teamspaces/projects/containers/containers.test.js:167]: similar sentiment here

File: backend/tests/v5/unit/processors/teamspaces/projects/federations/federations.test.js [hunk 13]
```diff
@@ -0,0 +1,143 @@
+/**
```
Comment (carmenfan) [inline backend/tests/v5/unit/processors/teamspaces/projects/federations/federations.test.js:1]: double license header here

Comment (carmenfan) [reply backend/tests/v5/unit/processors/teamspaces/projects/federations/federations.test.js:1]: again, similar comments to the containers tests

File: backend/tests/v5/e2e/routes/teamspaces/projects/containers/containers.test.js [hunk 14]
```diff
@@ -183,6 +183,39 @@ const testGetContainerStats = () => {
 	});
 };
 
+const testAppendFavourites = () => {
+	const route = `/v5/teamspaces/${teamspace}/projects/${project.id}/containers/favourites`;
+	describe('Append Favourite Containers', () => {
+		test('should fail without a valid session', async () => {
+			const res = await agent.get(route).expect(templates.notLoggedIn.status);
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/teamspaces/projects/containers/containers.test.js:190]: you're not testing your route?

this is a `HTTP get`, not a `HTTP patch` operation

**Review by carmenfan — CHANGES_REQUESTED** [review]

File: backend/src/v5/routes/teamspaces/projects/containers/containers.js [hunk 15]
```diff
@@ -43,6 +43,26 @@ const getContainerStats = async (req, res) => {
 	}).catch((err) => respond(req, res, err));
 };
 
+const deleteFavourites = (req, res) => {
+	const user = req.session.user.username;
+	const { teamspace, project } = req.params;
+	const favouritesToRemove = req.body.containers;
+
+	Containers.deleteFavourites(user, teamspace, project, favouritesToRemove).then((favourites) => {
+		respond(req, res, templates.ok, { favourites });
```
[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/src/v5/routes/teamspaces/projects/containers/containers.js:52]: For some reason if I don't pass a body to the respond method I never get a response in postman, so it seems that body has to be at least an empty object. I can either pass { } or have the param to have a default value of { }.

[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/routes/teamspaces/projects/containers/containers.js:52]: you mean it hangs or it returns with 200 just no body?

[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/routes/teamspaces/projects/containers/containers.js:52]: this probably points to a bug in responder, i'll take a look.

[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/routes/teamspaces/projects/containers/containers.js:52]: @ChristosTsiotsias  just commited a fix, does it work?

[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/src/v5/routes/teamspaces/projects/containers/containers.js:52]: yes seems to work now !

File: backend/src/v5/processors/teamspaces/projects/containers/containers.js [hunk 16]
```diff
@@ -67,4 +68,22 @@ Containers.getContainerStats = async (teamspace, project, container) => {
 	};
 };
 
+Containers.appendFavourites = async (username, teamspace, project, favouritesToAdd) => {
+	const accessibleContainers = await Containers.getContainerList(teamspace, project, username);
+	const containerIDs = accessibleContainers.map((a) => a._id);
```
[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/src/v5/processors/teamspaces/projects/containers/containers.js:73]: so for example something like that ?

```
{
    1: "0b680c90-dd07-11eb-92d6-8fe463212623",
    2: " "0b6a5680-dd07-11eb-92d6-8fe4632126231"
}
```

[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/processors/teamspaces/projects/containers/containers.js:73]: @ChristosTsiotsias  you probably want the id to be the key? more like
````js
{"0b680c90-dd07-11eb-92d6-8fe463212623": true, //... }
````

So at line 75 you just need to do `objectName[i]` to know if it exists

File: .vs/3drepo.io/config/applicationhost.config [hunk 17]
```diff
@@ -0,0 +1,958 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Charence) [inline .vs/3drepo.io/config/applicationhost.config:1]: This looks like your Visual Studio config file?

**Review by carmenfan — DISMISSED** [review]

**Review by carmenfan — APPROVED** [review]
